### PR TITLE
feat(qwen3): execute mixed LoRA batches

### PR DESCRIPTION
**Pull request by NolanHo**
## Summary

This PR makes Qwen3 LoRA serving execute mixed base/LoRA requests in one scheduler, executor, and model batch instead of serializing by adapter.

Main changes:

- Add Qwen3 LoRA capacity guardrails and server flags:
  - `--max-loras`
  - `--max-lora-rank`
- Set conservative built-in defaults for the supported knobs: `max_loras=1`, `max_lora_rank=64`.
- Carry per-request LoRA adapter metadata through scheduler plans, worker commands, executor calls, and Qwen3 forward paths.
- Remove the old global active-LoRA execution state and the tensor-parallel activation command.
- Run mixed-LoRA prefill, decode, and unified batches as one batch instead of one sub-batch per adapter.
- Add packed decode LoRA slots, per-token slot ids, grouped decode LoRA delta kernels, rank-specialized decode dispatch, and fixed-slot packed registry updates.
- Optimize rank-1 decode LoRA dot products with a block-wide reduction.
- Optimize packed LoRA B-slot loading with one CUDA pack kernel per projection slot instead of thousands of row-by-row D2D copies.
- Compact repeated-adapter prefill/unified ranges so non-contiguous requests using the same adapter share one compact LoRA delta path.
- Cap scheduler admission at the decode scratch/CUDA Graph bucket capacity so oversized active decode batches queue instead of panicking.
- Clean up LoRA packed slots and registry state on failed load/reload so packed decode state cannot diverge from the adapter registry.

## Behavior

Before this PR, the scheduler could admit requests using different adapters into one plan, but execution grouped by adapter and ran serial model forwards under a global active adapter. That preserved correctness but lost the main batching benefit for mixed base/LoRA traffic.

After this PR:

- The base model batch runs once for mixed base/LoRA traffic.
- Each request carries its adapter identity, or base/no-LoRA.
- Decode LoRA deltas are applied from packed adapter slots using per-token slot ids.
- Prefill/unified LoRA deltas are applied by adapter-grouped compact token ranges.
- No-LoRA decode still uses the existing CUDA Graph path.
- LoRA decode remains eager in this PR because packed LoRA slot versions are not part of the CUDA Graph cache key yet.

### Mixed-LoRA serving

Environment:

- GPU: `NVIDIA A800-SXM4-80GB`
- Model: Qwen3-4B
- API: HTTP `/v1/completions`
- Workload: `concurrency=12`, `rounds=5`, `60` measured requests, `max_tokens=16`
- Model mix after warmup: `qwen3-base`, `stress-a`, `stress-b`, `stress-c`

| Stage | Throughput | Mean latency | p50 | p90 | p99 |
| --- | ---: | ---: | ---: | ---: | ---: |
| Capacity baseline | `19.88 req/s` | `0.602s` | `0.601s` | `0.611s` | `0.613s` |
| Mixed-LoRA batch | `44.13 req/s` | `0.271s` | `0.270s` | `0.280s` | `0.281s` |

Delta:

- Throughput: `2.22x`
- Mean latency: `-54.94%`
- p50 latency: `-54.99%`
- p99 latency: `-54.15%`

### Rank-1 decode dot optimization

Environment is the same A800 workload as above.

| Stage | Throughput | Mean latency | p50 | p99 |
| --- | ---: | ---: | ---: | ---: |
| Mixed-LoRA batch | `44.13 req/s` | `0.271s` | `0.270s` | `0.281s` |
| Rank-1 block reduction | `60.94 req/s` | `0.196s` | `0.195s` | `0.203s` |

Delta vs mixed-LoRA batch:

- Throughput: `1.38x`
- Mean latency: `-27.67%`
- p50 latency: `-27.74%`
- p99 latency: `-27.89%`

Delta vs capacity baseline:

- Throughput: `3.07x`
- Mean latency: `-67.41%`
- p50 latency: `-67.48%`
- p99 latency: `-66.93%`

### Rank sweep after non-rank1 dot optimization

I also benchmarked a single-adapter q/v LoRA workload across common ranks after extending the decode dot path beyond rank 1.

Environment:

- GPU: `NVIDIA A800-SXM4-80GB`
- Model: Qwen3-4B
- API: HTTP `/v1/completions`
- Workload: one loaded q/v LoRA adapter, `concurrency=12`, `rounds=10`, `120` measured requests, `max_tokens=16`

| Rank | Throughput | Mean latency | p50 | p99 | Adapter load |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | `63.73 req/s` | `0.188s` | `0.188s` | `0.191s` | `0.0107s` |
| 2 | `60.13 req/s` | `0.199s` | `0.199s` | `0.202s` | `0.0127s` |
| 4 | `58.80 req/s` | `0.204s` | `0.204s` | `0.207s` | `0.0174s` |
| 8 | `52.36 req/s` | `0.229s` | `0.229s` | `0.232s` | `0.0181s` |
| 16 | `43.64 req/s` | `0.275s` | `0.275s` | `0.278s` | `0.0271s` |
| 32 | `32.75 req/s` | `0.366s` | `0.366s` | `0.369s` | `0.0472s` |
| 64 | `21.77 req/s` | `0.551s` | `0.551s` | `0.554s` | `0.0901s` |

In a same-workload historical comparison against the pre-fix non-rank1 decode dot path, this change improved:

- rank 8: `+45.2%` throughput
- rank 16: `+40.0%` throughput
- rank 32: `+34.4%` throughput
- rank 64: `+5.0%` throughput

Rank 64 still scales down materially because more of the cost shifts to the LoRA B expansion and total decode work.

### Adapter load optimization

The mixed path originally made adapter load slower because packed LoRA slots were populated at load time. This PR fixes that with a CUDA B-slot packing kernel.

Environment:

- GPU: `NVIDIA A800-SXM4-80GB`
- Model: Qwen3-4B
- API/workload: same mixed base/LoRA HTTP shape as above

Forward order:

| Stage | Adapter load mean | Throughput | Mean latency | p99 |
| --- | ---: | ---: | ---: | ---: |
| Before pack-B load kernel | `0.503s` | `56.04 req/s` | `0.213s` | `0.225s` |
| After pack-B load kernel | `0.00669s` | `56.70 req/s` | `0.210s` | `0.223s` |

Reverse order:

| Stage | Adapter load mean | Throughput | Mean latency | p99 |
| --- | ---: | ---: | ---: | ---: |
| After pack-B load kernel | `0.00671s` | `55.98 req/s` | `0.213s` | `0.227s` |
| Before pack-B load kernel | `0.490s` | `56.23 req/s` | `0.212s` | `0.225s` |

Conclusion:

- Adapter load improves from roughly `0.49-0.50s` to millisecond-scale loads; current-head reruns measured `~0.010-0.012s` for rank-1/mixed adapters, with the original forward/reverse comparison measuring `0.0067s`.
- Steady serving throughput after adapters are loaded is unchanged within benchmark noise for this workload.

### Steady-state nsys result

I also captured a steady-state mixed base/LoRA HTTP nsys trace before deciding whether to continue narrow LoRA kernel fusion.

These percentages come from one steady-state nsys profile over the measured request phase, not a general model-wide attribution.

Workload:

- Requests: `240`
- Concurrency: `12`
- `max_tokens=16`
- Models: `qwen3-base`, `stress-a`, `stress-b`, `stress-c`

Profiled request phase:

- Throughput: `58.98 req/s`
- Mean latency: `0.203s`
- p99 latency: `0.221s`

GPU kernel time composition:

- Base GEMM/cuBLAS-like projection kernels: `~67.8%`
- `argmax_kernel`: `15.9%`
- `lora_decode_fused_delta_group3_rank_kernel<1>`: `6.8%`
- Decode attention FlashInfer: `1.9%`
- Append paged KV: `1.0%`
- QK norm/RoPE: `1.4%`
- Fused add RMSNorm: `2.4%`
- `silu_mul`: `1.3%`

CUDA API time composition:

- Kernel launch APIs: `~27.8%`
- `cuMemcpyDtoHAsync_v2`: `65.8%`
- `cuStreamSynchronize`: `0.2%`

Follow-up priority from profiling:

- LoRA decode is `6.8%` of GPU kernel time in this measured steady-state workload.
- The next serving work should prioritize sampling/logits selection and D2H readback, then the base GEMM/MLP decode path.
- Revisit LoRA delta fusion for workloads with more target modules or higher ranks.

## Validation

Formatting and static checks:

```bash
cargo fmt --all --check
git diff --check
python -m py_compile tools/lora/qwen3_lora_live_parity.py tools/lora/qwen3_lora_live_stress.py
rg -n "activate_lora_adapter|active_lora_adapter|ActivateLoraAdapter|lora_layer\\(|apply_lora_projection_delta\\(" pegainfer-qwen3-4b pegainfer-core pegainfer-server pegainfer-kernels
```

The residual scan for the deleted global activation path returned no matches.

Build and unit tests run:

```bash
PEGAINFER_CUDA_SM=80 cargo check --release -p pegainfer-server --bin pegainfer
PEGAINFER_CUDA_SM=80 cargo check --release -p pegainfer-qwen3-4b --lib
PEGAINFER_CUDA_SM=80 cargo test --release -p pegainfer-qwen3-4b --lib -- --nocapture
PEGAINFER_CUDA_SM=80 cargo test --release -p pegainfer-qwen3-4b --lib lora -- --nocapture
PEGAINFER_CUDA_SM=80 cargo test --release -p pegainfer-qwen3-4b --lib admission -- --nocapture
PEGAINFER_CUDA_SM=80 cargo test --release -p pegainfer-qwen3-4b --lib packed_lora_registry_keeps_fixed_slots -- --nocapture
PEGAINFER_CUDA_SM=80 cargo test --release -p pegainfer-qwen3-4b --lib groups_non_contiguous_lora_ranges_by_adapter -- --nocapture
PEGAINFER_CUDA_SM=80 cargo test --release -p pegainfer-server --bin pegainfer lora -- --nocapture
PEGAINFER_CUDA_SM=80 cargo test --release -p pegainfer-qwen3-4b --test lora_smoke qwen3_lora_loads_rank64_adapter_and_generates -- --ignored --nocapture
```

Model-required validation:

`PEGAINFER_TEST_MODEL_PATH=<Qwen3-4B weights> cargo test --release -p pegainfer-qwen3-4b --test hf_golden_gate -- --nocapture` passed on A800:

| Gate pass | Positions | Head deltas | Mean | p50 | p99 | Max |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| sequential bs=1 eager | `816` | `6528` | `0.0318` | `0.0248` | `0.1180` | `0.3749` |
| batched eager (9, no pad) | `153` | `1224` | `0.0333` | `0.0257` | `0.1280` | `0.3749` |
| sequential bs=1 eager cached replay | `816` | `6528` | `0.0312` | `0.0246` | `0.1183` | `0.3749` |
| batched eager cached replay (9) | `153` | `1224` | `0.0308` | `0.0245` | `0.1250` | `0.1625` |
| batched cuda-graph (9 padded) | `153` | `1224` | `0.0300` | `0.0223` | `0.1281` | `0.1774` |
| batched cuda-graph (5 padded) | `85` | `680` | `0.0309` | `0.0262` | `0.1116` | `0.1777` |
| batched cuda-graph cached replay (5) | `85` | `680` | `0.0310` | `0.0240` | `0.1142` | `0.1826` |

Final result: `1 passed; 0 failed; finished in 32.53s`.

GPU runtime validation:

- On A800, `PEGAINFER_CUDA_SM=80 cargo test --release -p pegainfer-qwen3-4b --lib -- --nocapture` passed: `27` tests.
- With the shared Qwen3-4B model and CUDA compat libraries, `batch_decode::tests::batch_matches_sequential` passed.
- With the same model/runtime env, `cargo test --release -p pegainfer-qwen3-4b --test lora_smoke -- --ignored --nocapture` passed.
- A rank-64 zero LoRA adapter load + smoke generation test passed with the new default `max_lora_rank=64`.
- Rank sweep serving benchmarks completed for ranks `1/2/4/8/16/32/64`.
- Current-head mixed and rank-sweep HTTP benchmarks were re-run on two A800 validation hosts and matched the reported throughput/latency range.
- Mixed base/LoRA HTTP benchmarks completed on A800.
- Steady-state nsys profiling completed on A800.

## Known Boundaries / Follow-ups

- PegaInfer currently maps `max_loras` to resident loaded adapter capacity because there is no CPU adapter cache yet.
- LoRA decode remains eager. No-LoRA decode still uses the existing CUDA Graph path.
- LoRA CUDA Graph support would need packed-slot versioning in graph cache keys, otherwise adapter load/replacement can invalidate captured assumptions.
- Decode now has packed/grouped/rank-specialized LoRA delta kernels, but this is not full Punica parity.
- Based on the nsys trace, the next meaningful serving optimizations should target:
  - sampling/logits selection and D2H readback;
  - base GEMM/MLP decode projection cost;
  - only then revisit LoRA delta fusion for workloads with more target modules or higher ranks.

### CODE REVIEW

File: pegainfer-kernels/csrc/lora_fused.cu [hunk 1]
```diff
@@ -0,0 +1,790 @@
...
+    scratch[0] = value;
+  }
+  __syncthreads();
+  return scratch[0];
+}
+
+static size_t lora_rank_smem_bytes(int rank) {
+  int floats = rank == 1 ? 32 : rank;
+  return static_cast<size_t>(floats) * sizeof(float);
+}
```
[the remark this answers is not in the document]
Comment (NolanHo) [reply pegainfer-kernels/csrc/lora_fused.cu:31]: Fixed in 5cc9971: lora_rank_smem_bytes now allocates at least 32 floats for every rank, so block_reduce_sum scratch has enough space for all small-rank mixed/grouped paths.

File: pegainfer-core/src/ops/traced.rs [hunk 2]
```diff
@@ -86,6 +86,25 @@ pub fn gemm_into(
...
+    out: &mut HiddenStates,
+) -> Result<()> {
+    if call_trace::is_enabled() {
+        let label = call_trace::current_label("gemm_token_range");
+        call_trace::record_call(gemm_call::<OutDim, InDim>(
+            label,
+            weight.rows,
+            weight.cols,
+            out.seq_len,
+        ));
```
[the remark this answers is not in the document]
Comment (NolanHo) [reply pegainfer-core/src/ops/traced.rs:103]: Verified this one as a false positive. OutDim/InDim are imported marker types from pegainfer_kernels::tensor in traced.rs, and `PEGAINFER_CUDA_SM=80 cargo check --release -p pegainfer-core --features kernel-call-trace` passes on this branch.

File: pegainfer-qwen3-4b/src/lora.rs [hunk 3]
```diff
@@ -214,26 +330,80 @@ pub(crate) fn load_device_lora_adapter(
...
-    let mut rank_out = HiddenStates::zeros(ctx, projection.a.rows, input.seq_len)?;
-    ops::gemm_into(ctx, &projection.a, input, &mut rank_out);
-    let mut delta = HiddenStates::zeros(ctx, projection.b.rows, input.seq_len)?;
+    if token_indices.is_empty() {
+        return Ok(());
+    }
+    let token_indices_d = ctx
+        .stream
+        .clone_htod(token_indices)
+        .map_err(|e| anyhow::anyhow!("LoRA indexed token copy failed: {e}"))?;
```
[the remark this answers is not in the document]
Comment (NolanHo) [reply pegainfer-qwen3-4b/src/lora.rs]: Fixed in 5cc9971: indexed LoRA token groups are now prepared once per forward path and reused across projection applications/layers. apply_lora_projection_delta_indexed takes the prepared device token index slice directly, removing the per-projection H2D copy/allocation.

**Comment by xiaguan** [comment]
Wow — thanks for this PR. The rigor here is genuinely impressive: forward/reverse load-order ablation, p99 in every table, and especially using the nsys trace to decide *against* further LoRA fusion — that's exactly the right call. The per-request adapter identity + packed decode slots design is the right architecture, and deleting the adapter-grouping machinery in `plan.rs` is a real simplification. 2.2–3× on mixed traffic is a great result.

A few notes:

**1. Validation: `--test e2e` doesn't exist on this branch.** I suspect that log came from an older checkout — the exact-text e2e was removed in #194; the current accuracy gate is `tests/hf_golden_gate.rs`. Running that on this branch should also clear up the "Tell me a story" fixture confusion you hit. Could you paste its numbers?

**2. The eager fallback in `batch_decode` is dead code — deleting it makes the PR smaller *and* faster.** `lora_slots` is `Some` exactly when any request carries an adapter, which is exactly when `lora_groups` is non-empty. So the `!use_fused_lora` branches in `apply_decode_lora_projection*` can never execute a non-empty loop, and `prepare_lora_token_groups` does a device alloc + HtoD **every decode step** for interleaved batches (`[base, A, base, A]` — the common case now that the scheduler no longer groups by adapter), then the result is never used. Suggest: drop `lora_ranges`/`lora_groups` from `batch_decode` entirely, keep `lora_slots`, and gate the graph on `lora_slots.is_none()`.

**3. One-line insurance in `lora_fused.cu`:** `WARPS_PER_BLOCK = 8` is only correct because every launcher happens to use `block(256)`. Deriving it from `blockDim.x / warpSize` avoids silently skipping rank rows (and consuming a stale `rank_buf`) if anyone changes the block size later.

Follow-up ideas — non-blocking, but I'd love to see these as next PRs:

- **Flip the CUDA-graph kill switch for LoRA mode** (`main.rs`: `Qwen3 if enable_lora => false`, predates this PR). Your design already makes graphs safe: LoRA batches run eager and captures never contain LoRA kernels. Removing the switch gives base traffic in a LoRA-enabled server its graph speedup back — low-risk and very measurable.
- Packed-slot versioning in the graph cache key for LoRA-in-graph decode — your analysis in the description is exactly right.
- Run prefill off the packed slots too: that removes the double-resident adapter weights (registry copy + packed copy) and paves the way for the CPU adapter cache / `max_loras > 1` you mentioned.
- Your own nsys shows the group3 kernel at 6.8% GPU time with `grid = batch` and 3 projections serialized — `grid.y = 3` is a free 3× parallelism there.
- The rank-specialized kernels could collapse to roughly half the lines with a `RANK = 0 → runtime-rank` template trick — worth doing before Punica-parity work grows that file further.

Thanks again — this moves Qwen3 LoRA serving forward a lot.

**Comment by NolanHo** [comment]
Thanks for the careful review.

I rebased this onto current `upstream/main` and resolved the scheduler conflict by keeping both admission guards: context-window overflow still rejects, and decode-batch capacity pressure still defers instead of admitting an unsafe batch.

I also fixed the small post-rebase `batch_decode.rs` test-helper drift to the current `KvCacheManager::pool()` / `BlockPool` API.

**Review by xiaguan — APPROVED** [review]
LGTM
